crdb/retry.go

```diff
@@ -0,0 +1,284 @@
+// will cause the policy to fail early.
+type ExpBackoffRetryPolicy struct {
+	// RetryLimit is the maximum number of retries allowed. After this many
+	// retries, a MaxRetriesExceededError is returned.
+	RetryLimit int
+
+	// BaseDelay is the initial delay before the first retry. Each subsequent
+	// retry doubles this value: delay = BaseDelay * 2^(attempt-1).
+	BaseDelay time.Duration
+
+	// MaxDelay is the maximum delay cap. If > 0, delays are capped at this
+	// value once reached. If 0, delays grow unbounded (until overflow, which
+	// causes early termination).
+	MaxDelay time.Duration
+}
+
+// NewRetry implements RetryPolicy.
+func (l *ExpBackoffRetryPolicy) NewRetry() RetryFunc {
+	tryCount := 0
+	return func(err error) (time.Duration, error) {
+		tryCount++
+		if tryCount > l.RetryLimit {
+			return 0, newMaxRetriesExceededError(err, l.RetryLimit)
+		}
+		delay := l.BaseDelay << (tryCount - 1)
+		if l.MaxDelay > 0 && delay > l.MaxDelay {
+			return l.MaxDelay, nil
+		}
+		if delay < l.BaseDelay {
+			// We've overflowed.
+			if l.MaxDelay > 0 {
+				return l.MaxDelay, nil
+			}
+			// There's no max delay. Giving up is probably better in
+			// practice than using a 290-year MAX_INT delay.
+			return 0, newMaxRetriesExceededError(err, tryCount)
+		}
+		return delay, nil
+	}
+}
+
+// Vargo adapts third-party backoff strategies (like those from github.com/sethvargo/go-retry)
+// into a RetryPolicy without creating a direct dependency on those libraries.
+//
+// This function allows you to use any backoff implementation that conforms to the
+// VargoBackoff interface, providing flexibility to integrate external retry strategies
+// with CockroachDB transaction retries.
+//
+// Example usage with a hypothetical external backoff library:
+//
+//	import "github.com/sethvargo/go-retry"
+//
+//	// Create a retry policy using an external backoff strategy
+//	policy := crdb.Vargo(func() crdb.VargoBackoff {
+//	    // Fibonacci backoff: 1s, 1s, 2s, 3s, 5s, 8s...
+//	    return retry.NewFibonacci(1 * time.Second)
+//	})
+//	ctx := crdb.WithRetryPolicy(context.Background(), policy)
+//	err := crdb.ExecuteTx(ctx, db, nil, func(tx *sql.Tx) error {
+//	    // transaction logic
+//	})
+//
+// The function parameter should return a fresh VargoBackoff instance for each
+// transaction, as backoff state is not safe for concurrent use.
+func Vargo(fn func() VargoBackoff) RetryPolicy {
+	return &vargoAdapter{
+		DelegateFactory: fn,
+	}
+}
+
+// VargoBackoff is an interface for external backoff strategies that provide
+// delays through a Next() method. This allows adaptation of backoff policies
+// from libraries like github.com/sethvargo/go-retry without creating a direct
+// dependency.
+//
+// Next returns the next backoff duration and a boolean indicating whether to
+// stop retrying. When stop is true, the retry loop terminates with a
+// MaxRetriesExceededError.
+type VargoBackoff interface {
```

Comment:
And along with the above, let's rename this from `VargoBackoff` -> `ExternalBackoff`.